File: bff/src/services/orderService.ts
import axios from 'axios';
import { Order, OrdersResponse } from '@shared/dashboard';
import { logger } from '../utils/logger';

const ORDER_SERVICE_URL = process.env.ORDER_SERVICE_URL || 'http://localhost:3003';

export class OrderService {
    async getOrderCount(): Promise<number> {
        const response = await axios.get<{ count: number }>(
            `${ORDER_SERVICE_URL}/orders/count`
        );
        return response.data.count;
    }

    async getRecentOrders(limit: number = 5): Promise<Order[]> {
        const response = await axios.get<Order[]>(
            `${ORDER_SERVICE_URL}/orders/recent`,
            { params: { limit } }
        );
        return response.data;
    }

    async getRevenueData(days: number = 30): Promise<any[]> {
        const response = await axios.get<any[]>(
            `${ORDER_SERVICE_URL}/orders/revenue`,
            { params: { days } }
        );
        return response.data;
    }

    async getOrders(filters?: Record<string, any>): Promise<OrdersResponse> {
        const response = await axios.get<OrdersResponse>(
            `${ORDER_SERVICE_URL}/orders`,
            { params: filters }
        );
        return response.data;
    }

    async getOrderById(id: string): Promise<Order | null> {
        try {
            const response = await axios.get<Order>(
                `${ORDER_SERVICE_URL}/orders/${id}`
            );
            return response.data;
        } catch (error: any) {
            if (error.response?.status === 404) return null;
            logger.error(`Error fetching order ${id}:`, error);
            throw error;
        }
    }

    async updateOrderStatus(id: string, status: string): Promise<Order> {
        const response = await axios.patch<Order>(
            `${ORDER_SERVICE_URL}/orders/${id}/status`,
            { status }
        );
        return response.data;
    }
}

File: microservices/order-service/internal/service/order_service.go
package service

import (
	"github.com/samirllama/ecom-bff/order-service/internal/model"
	"github.com/samirllama/ecom-bff/order-service/internal/store"
)

type OrderService struct {
	store store.Store
}

func NewOrderService(s store.Store) *OrderService {
	return &OrderService{store: s}
}

func (s *OrderService) GetCount() int {
	c, err := s.store.GetOrderCount()
	if err != nil {
		return 0
	}
	return c
}

func (s *OrderService) GetRecentOrders(limit int) []model.Order {
	orders, err := s.store.GetRecentOrders(limit)
	if err != nil {
		return []model.Order{}
	}
	return orders
}

func (s *OrderService) GetRevenueData(days int) []model.RevenueItem {
	data, err := s.store.GetRevenueData(days)
	if err != nil {
		return []model.RevenueItem{}
	}
	return data
}

func (s *OrderService) GetOrders(page, limit int) *model.OrdersResponse {
	resp, err := s.store.GetOrders(page, limit, nil)
	if err != nil {
		return &model.OrdersResponse{Orders: []model.Order{}, Total: 0, Page: page, TotalPages: 0}
	}
	return resp
}

func (s *OrderService) GetOrderByID(id string) (model.Order, bool) {
	order, found, err := s.store.GetOrderByID(id)
	if err != nil || !found {
		return model.Order{}, false
	}
	return order, true
}

func (s *OrderService) UpdateOrderStatus(id, status string) (model.Order, error) {
	return s.store.UpdateOrderStatus(id, status)
}
